packages/crypto/src/bls12_318/pairing.rs

```diff
@@ -0,0 +1,171 @@
+use crate::CryptoError;
+
+use super::points::{g1_from_variable, g2_from_variable};
+use bls12_381::G2Prepared;
+use pairing::group::Group;
+use rayon::iter::{ParallelBridge, ParallelIterator};
+
+pub fn bls12_381_aggregate_pairing_equality(
+    ps: &[u8],
+    qs: &[u8],
+    r: &[u8],
+    s: &[u8],
+) -> Result<bool, CryptoError> {
+    let pq_pairs: Vec<_> = ps
+        .chunks_exact(48)
+        .zip(qs.chunks_exact(96))
+        // From here on parallelism is fine since the miller loop runs over
+        // a sum of the pairings and is therefore a commutative operation
+        .par_bridge()
+        .map(|(p, q)| {
+            let g1 = g1_from_variable(p)?;
+            let g2 = g2_from_variable(q)?;
+
+            Ok((g1.0, G2Prepared::from(g2.0)))
+        })
+        .collect::<Result<_, CryptoError>>()?;
+
+    let r = g1_from_variable(r)?;
+    let s = g2_from_variable(s)?;
+
+    let r_neg = -r.0;
+    let s_prepared = G2Prepared::from(s.0);
+
+    let ref_pq: Vec<_> = pq_pairs
+        .iter()
+        .map(|(g1, g2)| (g1, g2))
```

Comment:
What's this map doing?

Reply:
Converting a `&(G1, G2)` into a `(&G1, &G2)`, so not exactly an identity function unfortunately

Reply:
I see, thanks